### DISCUSSION TITLE
Setup monorepo prettier formatting on save in VSCode

### DISCUSSION THREAD

**Discussion by remy90** [vercel/turborepo#4104] [Help]
I'm trying to format an edited file when saving in VSCode for the Basic example template.
If i run `pnpm run format`, prettier does its magic against all files across all workspaces.

My `.vscode/settings.json` looks like this:

 ```json
{
  "eslint.packageManager": "pnpm",
  "eslint.workingDirectories": [
    "apps",
    "packages"
  ],
  "editor.defaultFormatter": "esbenp.prettier-vscode",
  "prettier.printWidth": 100,
  "prettier.arrowParens": "avoid",
  "prettier.enable": true,
  "prettier.endOfLine": "lf",
  "prettier.jsxSingleQuote": false,
  "editor.formatOnSave": true,
  "editor.codeActionsOnSave": {
    "source.fixAll.eslint": true,
    "source.organizeImports": true
  },
}
```

I get the feeling this is potentially an issue with prettier-vscode and I've [raised an issue](https://github.com/prettier/prettier-vscode/issues/2895) but wanted to create a discussion here, as I imagine I may not be the only person interested in tracking this issue.

**Comment by remy90** [comment]
Managed to get to the bottom of my issue.

For some reason `"editor.defaultFormatter": "esbenp.prettier-vscode",`. was insufficient.
I had manual entries per language to get this to work:
```json
"[javascript]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "[javascriptreact]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "[typescript]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "[typescriptreact]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
```

I would also recommend Monorepo Workspace extension.

**Reply by lukeocodes** [reply]
I'm worried this is the accepted answer as the official resolution for this

**Reply by ansh** [reply]
Same Luke

**Reply by lukeocodes** [reply]
> Same Luke

I got something working for my project now. When I get back to my desk, I will share my fix and it works brilliantly.

**Reply by anthonyshew** [reply]
I've unmarked that as the accepted answer as it smells funny to me, as well.

I've never been able to reproduce formatter issues in our starters and, when working with folks who have complaints around them not working, it typically ends up being a userland configuration issue, usually in the IDE.

**Reply by lari-ks** [reply]
The answer given above is actually correct though, and has been for several years.

Yes you need to set the default formatter for different languages in vscode.

Sometimes because a user will have multiple extensions that format more than one language, rather than necessarily configuring anything incorrectly.. so its recommended in documentation to be specific or to customise if needed and this has been the case for years. 

I have defaultFormatter specified for every language in every Vscode settings for every project, framework, workspace etc. There is no other settings or configurations in my Vscode aside from default. Only project specific.

Now 'workspace' as in monorepo / turborepo workspace is a different ballgame. I'm having some difficulty with that rn.
**Putting a .vscode on a monorepo root only doesn't really work, it does autosave (with defaultFormatters for each language)..... but ignores the config files for formatting and linting as they're in the repos, not the root of a turborepo .....** 

i'm trying to figure out vscode .code-workspace files so i can set settings for each repo ... and thats not working at all so far for me yet. I tried the monorepo extension too but no joy yet lol. 

But i do know defaultFormatter for each language will still be required in the settings.json in each prohect even if i make .code-workspaces work.

Doc:
https://code.visualstudio.com/docs/configure/settings#_language-specific-editor-settings

Example for Next app:

```
  "settings": {
      "editor.defaultFormatter": "esbenp.prettier-vscode",
      "editor.formatOnSave": true,
      "[json]": {
        "editor.defaultFormatter": "esbenp.prettier-vscode",
        "editor.formatOnSave": true
      },
      "[javascript]": {
        "editor.defaultFormatter": "esbenp.prettier-vscode"
      },
      "[typescript]": {
        "editor.defaultFormatter": "esbenp.prettier-vscode"
      },
      "[typescriptreact]": {
        "editor.defaultFormatter": "esbenp.prettier-vscode"
      }
  }
```

**Comment by servrox** [comment]
Using this now https://github.com/servrox/turbo-bun-tailwind-starter/blob/main/.vscode/settings.json
